pipeline/customer/customer_test.py

```diff
@@ -0,0 +1,397 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from decimal import Decimal
+import unittest
+
+import apache_beam as beam
+
+from pipeline.customer import customer, detail, order
+from pipeline.customer.customer import Validate, JsonFormatter
+
+
+class TestValidate(unittest.TestCase):
+
+    def test_empty(self):
+        data = (3, {
+            customer.Field.Detail: [],
+            customer.Field.Orders: [],
+        })
+        expected = [(3, {
+            customer.Field.CustomerId: 3,
+            customer.Field.Detail: None,
+            customer.Field.Orders: None,
+            customer.Field.Errors: ['no customer details'],
+        })]
+        actual = [data] | beam.ParDo(Validate())
+        self.assertEqual(expected, actual)
+
+    def test_no_details(self):
+        data = (3, {
+            customer.Field.Detail: [],
+            customer.Field.Orders: [{
+                order.Field.CustomerID: 3,
+                order.Field.Orders: [
+                    {
+                        order.Field.Id: 3607,
+                        order.Field.CustomerID: 3,
+                        order.Field.TotalPrice: Decimal('373.02'),
+                    },
+                    {
+                        order.Field.Id: 3949,
+                        order.Field.CustomerID: 3,
+                        order.Field.TotalPrice: Decimal('702.88'),
+                    },
+                ],
+                order.Field.Min: Decimal('373.02'),
+                order.Field.Max: Decimal('702.88'),
+                order.Field.Sum: Decimal('1075.90'),
+                order.Field.Avg: Decimal('537.95'),
+                order.Field.Error: [],
+            }],
+        })
+        expected = [(3, {
+            customer.Field.CustomerId: 3,
+            customer.Field.Detail: None,
+            customer.Field.Orders: {
+                order.Field.CustomerID: 3,
+                order.Field.Orders: [
+                    {
+                        'customer_id': 3,
+                        'id': 3607,
+                        'total_price': Decimal('373.02'),
+                    },
+                    {
+                        'customer_id': 3,
+                        'id': 3949,
+                        'total_price': Decimal('702.88'),
+                    }],
+                order.Field.Min: Decimal('373.02'),
+                order.Field.Max: Decimal('702.88'),
+                order.Field.Sum: Decimal('1075.90'),
+                order.Field.Avg: Decimal('537.95'),
+                order.Field.Error: [],
+            },
+            customer.Field.Errors: [u'no customer details'],
```

Comment:
The `u'` (unicode) is not needed because of the `from __future__ import unicode_literals`.

Reply:
done